### PR TITLE
Adding EIP-1559 Support for Tx Construction

### PR DESCRIPTION
**Pull request by ash-krishnan**
With EIP-1559, transaction fees are calculated using (baseFee + priority tip) * gas limit instead of gas price * gas limit.
In metadata, we fetch base fee from the latest block. And we use our new client function  `SuggestGasTip` which calls RPC endpoint `eth_maxPriorityFeePerGas` to fetch the priority tip. Gas limit stays unchanged. We set gas price to zero because we don't need it to construct dynamic fee transactions and we are dropping support of constructing legacy transactions. But we still need gas price in the data API because we still need to be able to read/parse legacy transactions.

Gas Cap can be overridden with user input but by default it is (baseFee * 2 + priority tip). 

We have also changed the signer we previously use (EIP-155Signer, which only supports legacy transactions) to the LondonSigner, which supports dynamic fee transactions.

In SendTransaction, dynamic fee transactions will only successfully broadcast if we remove the first two bytes of the transaction data, more information here: https://ethereum.stackexchange.com/questions/124447/eth-sendrawtransaction-with-dynamicfeetx-returns-expected-input-list-for-types

Tested transactions on Mumbai testnet:
Native: 0xba567863128cd87ec3e97afede82ab6746a93342c21d097a4fb632edc7dab567
ERC20: 0x22582aab59072c964fe2669450903392fbd464a03543d2dbf2ef96272f61d647

### CODE REVIEW

File: services/construction/types.go [hunk 1]
```diff
@@ -364,7 +427,7 @@ func (t *transaction) UnmarshalJSON(data []byte) error {
 		return err
 	}
 
-	gasPrice, err := hexutil.DecodeBig(tw.GasPrice)
+	gasPrice := big.NewInt(0)
```
Comment (andrewkmin) [inline services/construction/types.go:430]: qq: any particular reason we're manually setting `gasPrice` to 0 here?

File: polygon/client.go [hunk 2]
```diff
@@ -255,7 +277,9 @@ func (ec *Client) SendTransaction(ctx context.Context, tx *types.Transaction) er
 	if err != nil {
 		return err
 	}
-	return ec.c.CallContext(ctx, nil, "eth_sendRawTransaction", hexutil.Encode(data))
+	// We have to remove the first two bytes otherwise DynamicFeeTxs will not send:
```
Comment (andrewkmin) [inline polygon/client.go:280]: Thanks for this context 🙏

File: polygon/client.go [hunk 3]
```diff
@@ -193,6 +194,27 @@ func (ec *Client) PendingNonceAt(ctx context.Context, account common.Address) (u
 	return uint64(result), err
 }
 
+// SuggestGasTipCap retrieves the currently suggested gas tip cap after 1559 to
+// allow a timely execution of a transaction.
+//func (ec *Client) GetFeeHistory(ctx context.Context) (*big.Int, error) {
```
Comment (andrewkmin) [inline polygon/client.go:199]: Maybe we can table this as a follow up item/alternative method if we need more "up-to-date" estimates

File: services/construction/metadata.go [hunk 4]
```diff
@@ -108,16 +107,35 @@ func (a *APIService) ConstructionMetadata(
...
 
+	var gasCap *big.Int
+	if input.GasCap == nil {
+		multiplier := big.NewInt(2)
+		gasCap = new(big.Int).Add(gasTip, new(big.Int).Mul(header.BaseFee, multiplier))
+	} else {
+		gasCap = input.GasCap
+	}
+
+	// gasPrice, err := a.client.SuggestGasPrice(ctx, input.GasPrice)
```
Comment (andrewkmin) [inline services/construction/metadata.go:128]: Do we want to use the live estimate here?

File: services/construction/types.go (hunk 1 above)
Comment (ash-krishnan) [reply services/construction/types.go:430]: GasPrice isn't used by Dynamic Fee txs. So we can get the live estimate and set in the unsigned transaction. But in `combine.go` we can't set Gas Price in the dynamic tx fee struct: (https://github.com/maticnetwork/polygon-rosetta/blob/7075fdf80566b8b2a3328b47a3d56b1f19d1efbf/services/construction/combine.go#L46). So when the signed transaction has no gas price. We also don't use it in our calculations so we're also making extra RPC calls for information we're not using.

File: polygon/client.go (hunk 3 above)
Comment (ash-krishnan) [reply polygon/client.go:199]: Yup, forgot to remove this, I'll get rid of it

File: services/construction/metadata.go (hunk 4 above)
Comment (ash-krishnan) [reply services/construction/metadata.go:128]: We're not really using gas price anymore in the construction API so getting live estimates might be unnecessary here. We also can't set gas price in the dynamic fee tx in `combine.go`.

File: services/construction/types.go [hunk 5]
```diff
@@ -324,6 +382,8 @@ type transactionWire struct {
 	Data     string `json:"data"`
 	Nonce    string `json:"nonce"`
 	GasPrice string `json:"gas_price"`
```
Comment (roberto-bayardo) [inline services/construction/types.go:384]: Can this be removed entirely now that we don't construct old type txs?

File: services/construction/preprocess.go [hunk 6]
```diff
@@ -95,6 +95,25 @@ func (a *APIService) ConstructionPreprocess(
...
+		if !ok {
+			return nil, svcErrors.WrapErr(
+				svcErrors.ErrInvalidGasCap,
+				fmt.Errorf("%s is not a valid gas_cap", v),
+			)
+		}
+		preprocessOutputOptions.GasCap = bigObj
+	}
+
 	// Override gas_price
```
Comment (roberto-bayardo) [inline services/construction/preprocess.go:117]: can we get rid of this? Overriding gas price shouldn't do anything if we construct only 1559 type transactions right?

File: services/construction/types.go [hunk 7]
```diff
@@ -80,6 +84,8 @@ type options struct {
 	Value           *big.Int `json:"value,omitempty"`
 	GasPrice        *big.Int `json:"gas_price,omitempty"`
```
Comment (roberto-bayardo) [inline services/construction/types.go:85]: can we get rid of this?

File: services/construction/types.go [hunk 8]
```diff
@@ -94,6 +100,8 @@ type optionsWire struct {
 	Value           string   `json:"value,omitempty"`
 	GasPrice        string   `json:"gas_price,omitempty"`
```
Comment (roberto-bayardo) [inline services/construction/types.go:101]: same .. get rid of this if never going to be used anymore?

**Review by roberto-bayardo — CHANGES_REQUESTED** [review]
Most suggestions are around blowing away more of the legacy GasPrice code (if it's possible, which it seems it should be)

File: services/construction/types.go (hunk 5 above)
Comment (ash-krishnan) [reply services/construction/types.go:382]: like it looks like gas price is still being used here? https://github.com/maticnetwork/polygon-rosetta/blob/c082aa44e76690a59c6fdc31868c058360e856e2/polygon/client.go#L376 or is the `transaction` struct just being used in construction

Comment (roberto-bayardo) [reply services/construction/types.go:382]: that gasPrice is just a local variable initialized by calling EffectiveGasPrice.   Yeah I was thinking the structs might just be part of construction. (I could be wrong though ...)

File: services/construction/types.go (hunk 7 above)
Comment (ash-krishnan) [reply services/construction/types.go:83]: could we get a confirmation that removing `gas_price` in the `transaction` and `option` structs and generally from construction will not affect block/transaction parsing?

Comment (andrewkmin) [reply services/construction/types.go:83]: one scenario i could envision is if we ever want/need to manually pass in a super high gas price, likely to be used with a legacy transaction. this is what nova currently does for ETH's replacement transactions (use 1559 by default, but legacy for replacements). however, if that seems pointless, then we could remove `gas_price` here

the TLDR of the above ^ is: will EIP1559 ever put us into a place where we're unable to send a tx through with high urgency?

Comment (roberto-bayardo) [reply services/construction/types.go:83]: That can be accomplished by setting a very high miner tip.   I don't think there's any reason to be creating legacy transactions as 1559 is equal or better in every dimension.

Comment (ash-krishnan) [reply services/construction/types.go:83]: agreed we can create an override for the priority tip to be set in preprocess like we were doing for `gas_price`. in that case, should i remove `gas_price`?

**Comment by roberto-bayardo** [comment]
If it doesn't break anything I would highly recommend it!

File: services/construction/types.go (hunk 7 above)
Comment (roberto-bayardo) [reply services/construction/types.go:83]: If it doesn't break anything I would recommend it!  Ok w/ you @andrewkmin ?

Comment (ash-krishnan) [reply services/construction/types.go:83]: Cool, the `block` endpoint still works fine so I'll take it out

**Review by VAIBHAVJINDAL3012 — APPROVED** [review]

**Pull request merged**